Type inference plumbing

jlapacik opened the pull request:
### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

#1953 .

This PR does several things.

1. It defines empty structs for each of the basic types. Previously each of the basic types (int, float, string, ...) were just variants of the `MonoType` enum.

2. Defines the data structure for type substitutions as well as the `Substitutable` trait.

3. Defines the data structure for mapping type variables to their kind constraints.

4. Defines the `Constraint` enum which represents either a type constraint or a kind constraint. Type inference will produce a set of these constraints for each expression.

affo commented on libflux/src/semantic/infer.rs:
```diff
@@ -0,0 +1,49 @@
+            Constraint::Kind(MonoType::Var(Tvar(3)), Kind::Divisible),
+        ]);
+        assert_eq!(
+            c0 + c1,
```
Oh gosh, so cool. Summing up constraints

nathanielc commented on libflux/src/semantic/types.rs:
```diff
@@ -116,27 +209,490 @@ impl fmt::Display for MonoType {
+            MonoType::Bool(t) => {
+                if let MonoType::Var(tv) = with {
+                    tv.unify(self, cons)
+                } else if let MonoType::Bool(s) = with {
```
My reading of this code says that the:

```
if let MonoType::Var(tv) = with {
                    tv.unify(self, cons)
}
```

code can be moved out of the `match self` block. Maybe this is because of the `Self` reference in the `unifiable` trait again?

affo replied:
Yep, I agree.

@jlapacik I also think that two nested `match` expression would make the code more readable.

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -58,13 +49,102 @@ impl fmt::Display for PolyType {
+//                       ^_______________________^
+//
+#[derive(Debug, Clone, PartialEq)]
+pub struct Const {
```
Maybe? I have 3 different structs with some version of Constraint in their names, so I definitely need a better naming scheme.

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -58,13 +49,102 @@ impl fmt::Display for PolyType {
+// with the substitution equating the two types.
+//
+trait Unifiable {
+    fn unify(self, with: Self, cons: Const) -> Result;
```
Yes `Self` refers to the implementing type. So you could **not** pass an integer to the unify method of `Bool`.

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -75,22 +155,35 @@ impl fmt::Display for Kind {
 
+impl cmp::Ord for Kind {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.to_string().cmp(&other.to_string())
```
Yes there should be a comment here. This is for display purposes only.

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -75,22 +155,35 @@ impl fmt::Display for Kind {
+
+impl cmp::PartialOrd for Kind {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
```
Yes sorry for the confusion. Same reason as above. Display purposes only. Specifically displaying the kind constraints of a polytype's free variables.

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -58,13 +49,102 @@ impl fmt::Display for PolyType {
+#[derive(Debug)]
+enum Error {
+    Constrain { t: MonoType, with: Kind },
+    Unify { t: MonoType, with: MonoType },
```
Yes sorry this wasn't clear. I thought constrain was better verbiage for kinds since unify connotes equality more or less. I guess the doc comment for `Constrainable` didn't explain this very well?

I agree I have too many names with Constraint in them. I'll try to come up with more descriptive ones.

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -58,13 +49,102 @@ impl fmt::Display for PolyType {
+    fn constrain(self, with: Kind, cons: Const) -> Result;
+}
+
+type Result = result::Result<(Subst, Const), Error>;
```
Yes you're right this probably does need to be public.

Unification returns a substitution that is a unifier for two types. But when we unify types we may also discover partial information about type variables in the form of kind constraints. This is why we return both.

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -116,27 +209,490 @@ impl fmt::Display for MonoType {
+
+impl Unifiable for Bool {
+    fn unify(self, _: Self, cons: Const) -> Result {
+        Ok((Subst::empty(), cons))
```
You raise a good point. I guess this begs the question, does `Unifiable` make sense as a trait at all? Unlike `Substitutable` it's not implemented on any external types. Perhaps it's just confusing?

I also defined empty structs for the basic types like `Int`, `Float`, etc. I did this to isolate each type's unification logic. But does it make sense to have a `struct Bool;` type, or should there just be a single `MonoType::Bool` variant with the unification logic for `Bool` implemented as part of `MonoType::unify`?

nathanielc replied:
> I also defined empty structs for the basic types like Int, Float, etc. I did this to isolate each type's unification logic. But does it make sense to have a struct Bool; type, or should there just be a single MonoType::Bool variant with the unification logic for Bool implemented as part of MonoType::unify?

This is what I expected when I first saw the MonoType definition. I was confused at first why the basic types were being reintroduced as their own struct. 

Maybe try it out? Remove the trait and see if it simplifies the logic?

affo replied:
@jlapacik I like your proposal of just using variants and implement the logic directly into `Monotype`.

affo commented on libflux/src/semantic/types.rs:
```diff
@@ -240,82 +821,114 @@ pub struct Function {
+
+impl Unifiable for Function {
+    fn unify(self, _: Self, _: Const) -> Result {
+        unimplemented!();
```
those `unimplemented!` are left for future implementation in another PR?

jlapacik replied:
Yes correct.

jlapacik commented on libflux/src/semantic/infer.rs:
```diff
@@ -0,0 +1,49 @@
+}
+
+#[derive(Debug, PartialEq)]
+struct Constraints(Vec<Constraint>);
```
If it's just an alias to an external type, you can't implement any external traits on it. I wouldn't be able to implement `ops::Add` or `fmt::Display`.

affo replied:
That's a good point, I thought you could, but _an alias is just an alias_ 😅

jlapacik commented on libflux/src/semantic/sub.rs:
```diff
@@ -0,0 +1,55 @@
+// idempotent. Given a monotype x, and a substitution s, s(s(x)) = s(x).
+//
+#[derive(Debug)]
+pub struct Subst(HashMap<Tvar, MonoType>);
```
Same as above. I implemented the `fmt::Display` trait for substitutions. If it weren't for that implementation you are right, this could be just an alias. The Go code has methods for displaying substitutions and constraint sets. I found them to be quite useful when debugging.

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -58,13 +49,102 @@ impl fmt::Display for PolyType {
+            values: self
+                .kinds
+                .iter()
+                .collect::<BTreeMap<_, _>>()
```
By collecting into a BTree we obtain a sorted iterator. This is for consistent(vs random) displaying of output.

affo replied:
Ooooo, got it!

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -58,13 +49,102 @@ impl fmt::Display for PolyType {
+            "{}:{}",
+            self.tv,
+            DisplayList {
+                values: self.kinds.iter().collect::<BTreeSet<_>>().iter().collect(),
```
Same as above. A little trick to obtain sorted output.

affo replied:
Great! Can you put a comment, just to be explicit?

If you directly use that data structures you loose performance on insert/update right? From O(1) to O(logn).

jlapacik replied:
Absolutely. And yes insert is O(logn). I'm not too concerned with the performance characteristics of display though.

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -58,13 +49,102 @@ impl fmt::Display for PolyType {
+
+impl fmt::Display for Const {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        DisplayList {
```
I completely overlooked `join`. That's exactly what I was looking for. Thank you.

jlapacik commented:
@nathanielc @affo thank you for the reviews. I will implement the unification for type variables so that it's more clear how these methods interact with each other.

jlapacik commented:
I have updated the comments and removed unnecessary traits. I also removed the empty structs for basic types. All the unification logic for basic types is now part of `MonoType`. I also implemented unification for type variables in the last commit with some test cases.

affo commented on libflux/src/semantic/types.rs:
```diff
@@ -212,133 +191,71 @@ impl Substitutable for MonoType {
-                }
-            }
-            MonoType::Var(t) => t.unify(with, cons),
+        match (self, with) {
```
Boom combinatory `match`. Didn't know about this

nathanielc commented on libflux/src/semantic/types.rs:
```diff
@@ -116,27 +205,251 @@ impl fmt::Display for MonoType {
+            } else {
+                Err(Error::Constrain {
+                    t: self,
+                    with: k.difference(&kinds).map(|&kind| kind).collect(),
```
what does the `map` call do here? I looks like the identity as far as I can tell.

jlapacik replied:
Basically. The error enum owns the kinds passed into it, so I have to collect a set of kinds vs a set of references to kinds. The error enum could probably use a little more work.

nathanielc replied:
ok so it converts `[Kind]` to `[&Kind]`? Should we instead use `&[Kind]`?

jlapacik replied:
It converts `[&Kind]` to `[Kind]`. I think it makes sense for errors to consume their kinds. The pattern is that kinds are owned, they're not referenced, so the same should hold for errors.

nathanielc commented on libflux/src/semantic/types.rs:
```diff
@@ -116,27 +205,251 @@ impl fmt::Display for MonoType {
+        let kinds: HashSet<Kind> = cons
+            .0
+            .get(&self)
+            .unwrap_or(&HashSet::new())
```
nit, in reading up on what this does, I discovered that `unwrap_or` is evaluated eagerly so we are creating a new hashset with each call to `unify_with_tvar`. The docs recommend using `unwrap_or_else` which accepts a closure which is evaluated lazily.

jlapacik replied:
I must have missed `unwrap_or_else`. I will update. Note however that while this creates a new hashset value, it won't actually allocate any heap memory until we insert the first element.

nathanielc replied:
That is good to know, I see that in the docs now. In that case i say this reads more clearly so we should leave it as `unwrap_or` since we do not need to avoid the allocation.

nathanielc commented on libflux/src/semantic/types.rs:
```diff
@@ -116,27 +205,251 @@ impl fmt::Display for MonoType {
+            // Update the kind constraints
+            cons.0.remove(&self);
+            cons.0.remove(&tv);
+            cons.0.insert(tv, kinds);
```
insert will replace the existing value, why do we remove it first?

jlapacik replied:
Typo

affo commented on libflux/src/semantic/types.rs:
```diff
@@ -14,57 +19,138 @@ use std::{cmp, collections::HashMap, fmt};
+// structure.
+//
+#[derive(Debug, Clone, PartialEq)]
+pub struct Const(HashMap<Tvar, HashSet<Kind>>);
```
Why not `Constraints`? `Const` makes me think about "constant".

jlapacik replied:
There's already a `Constraints` struct for the constraints that inference generates. Maybe `TvarKinds`?

affo commented on libflux/src/semantic/types.rs:
```diff
@@ -116,9 +217,136 @@ impl fmt::Display for MonoType {
+                Kind::Comparable,
+                Kind::Equatable,
+                Kind::Nullable
+            ])(with, cons),
```
It seems like every time you constrain something a new hashset gets created. Could we store those constant hashsets in something generated once for all?

jlapacik replied:
This could be handled more efficiently using nested match expressions. This ways is nice because all of the constraint and error logic is in a single closure. Let's optimize this later.

affo commented on libflux/src/semantic/types.rs:
```diff
@@ -116,9 +217,136 @@ impl fmt::Display for MonoType {
+        }
+    }
+
+    fn constrain(self, with: HashSet<Kind>, cons: Const) -> Result {
```
Why do you need `cons: Const` here? Just to return a `Result`? 🤔

affo commented on libflux/src/semantic/types.rs:
```diff
@@ -141,16 +369,109 @@ impl Fresher {
+        }
+    }
+
+    fn constrain(self, with: HashSet<Kind>, cons: Const) -> Result {
```
Ok, I see. You need `Const` above because you need it here, right?

But there is no trait to implement, so... 🤔

jlapacik replied:
It's also needed for Array, Row, and Function.

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -81,19 +82,51 @@ fn display_kinds(tv: Tvar, kinds: &HashSet<Kind>) -> String {
+    fn update(self, tv: Tvar, with: HashSet<Kind>) -> Self {
+        let mut constraints = self;
+        if let Some(kinds) = constraints.0.get(&tv) {
+            let new: HashSet<Kind> = kinds.union(&with).map(|&kind| kind).collect();
```
Unfortunately no. I believe the compiler said it's missing the `From` trait.
